doc/articles/StructChaining.md

```diff
@@ -0,0 +1,81 @@
+
+Once the implementation identifies the struct by its `sType`, it casts the pointer back to the appropriate struct type in order to access its contents. Setting `sType` incorrectly (or pointing to any type that isn't a subtype of @ref WGPUChainedStruct) causes undefined behavior.
+
+```c
+typedef enum WGPUSType {
+    // ...
+    WGPUSType_MyStructExtension1 = /* ... */,
+    WGPUSType_MyStructExtension2 = /* ... */,
+    // ...
+} WGPUStype;
+
+typedef struct WGPUMyStructExtension1 {
+    WGPUChainedStruct chain; // .chain.sType must be WGPUSType_MyStructExtension1
+    uint32_t y;
+} WGPUMyStructExtension1;
+
+typedef struct WGPUMyStructExtension2 {
+    WGPUChainedStruct chain; // .chain.sType must be WGPUSType_MyStructExtension2
+    uint32_t z;
+} WGPUMyStructExtension2;
+```
+
+This is used like so:
+
+```c
+WGPUMyStructExtension2 ext2 = WGPU_MY_STRUCT_EXTENSION2_INIT;
+// .chain.sType is already set correctly by the INIT macro.
+// .chain.next is set to NULL indicating the end of the chain.
+ext2.z = 2;
+
+WGPUMyStructExtension1 ext1 = WGPU_MY_STRUCT_EXTENSION1_INIT;
+// .chain.sType is already set correctly by the INIT macro.
+// .chain.next may be set in either of two ways, equivalently:
+ext1.chain.next = &ext2.chain;
+ext1.chain.next = (WGPUChainedStruct*) &ext2;
+ext1.y = 1;
+
+WGPUMyStructBase base = WGPU_MY_STRUCT_BASE_INIT;
+// Note: base structs do not have STypes (they are statically typed).
+base.nextInChain = &ext1.chain;
+base.x = 0;
+```
+
+The pointer links in a struct chain are all mutable pointers. Whether the structs in the chain are actually mutated depends on the function they are passed to (whether the struct chain is passed as an input or an output). Some structs (e.g. @ref WGPULimits) may be either an input or an output depending on the function being called.
+
+## Struct-Chaining Error {#StructChainingError}
+
+A struct-chaining error occurs if a struct chain is incorrectly constructed (in a detectable way).
+They occur if and only if:
+
+- The `sType` of a struct in the chain is not valid for the type of the chain root.
+- Multiple instances of the same `sType` value are seen in the same struct chain. (Note this also detects and disallows cycles.)
+
+Struct chains which are used in device-timeline validation/operations (e.g. @ref WGPUBufferDescriptor in @ref wgpuDeviceCreateBuffer) have their chain errors surfaced asynchronously, like any other validation error.
+
+Struct chains which are used in content-timeline operations (e.g. @ref OutStructChainError) have their chain errors surfaced synchronously, like other content-timeline validation errors.
+
+### Out-Struct-Chain Error {#OutStructChainError}
+
+Operations which take out-struct-chains (e.g. @ref WGPULimits, in @ref wgpuAdapterGetLimits and @ref wgpuDeviceGetLimits, but not in @ref WGPUDeviceDescriptor) handle struct-chaining errors as follows:
+
+- The output struct and struct chain is not modified.
```

Comment:
Is this what we decided? I thought that we decided that this might be hard for implementation in cases of multiple output struct chains. i.e. It might be harder for implementations to handle resetting everything in the chain if a later chain errors?

My initial thought was that if error, users should still call `FreeMember` on all output structs.  For any struct that we successfully wrote to, that will free the resources, and on the ones we error-ed on, we will not write anything, and hence if users used INIT macros to initialize the out structs, `FreeMember` should work and not do anything.

Reply:
I added this in #377 (it just moved in this PR), but that was when we thought `FreeMembers` would be called on the whole chain, not on individual structs in the chain. Then it wouldn't make any sense to tell people to call `FreeMembers` because `FreeMembers` wouldn't be able to traverse the chain either. (Which now that I think about it is another reason `FreeMembers` perhaps shouldn't traverse the chain; it would need to return a `wgpu::Status`.)

That said I think the only thing implementations need to do to implement this correctly is unpack the struct chain before writing anything to it. This seems kinda necessary anyway because if the struct hasn't been unpacked we can't even really write anything out except for the base struct.

Reply:
Ack!